src/circuits/etable_compact/op_configure/op_unary.rs

```diff
@@ -0,0 +1,437 @@
+use super::*;
+use crate::{
+    circuits::{
+        mtable_compact::encode::MemoryTableLookupEncode,
+        rtable::pow_table_encode,
+        utils::{bn_to_field, Context},
+    },
+    constant,
+};
+use halo2_proofs::{
+    arithmetic::FieldExt,
+    plonk::{Error, Expression, VirtualCells},
+};
+use specs::{
+    etable::EventTableEntry,
+    itable::{OpcodeClass, UnaryOp, OPCODE_ARG0_SHIFT, OPCODE_ARG1_SHIFT, OPCODE_CLASS_SHIFT},
+    mtable::VarType,
+};
+
+pub struct UnaryConfig {
+    operand: U64Cell,
+    result: U64Cell,
+    operand_inv: UnlimitedCell,
+    operand_is_zero: BitCell,
+
+    is_ctz: BitCell,
+    is_clz: BitCell,
+    is_popcnt: BitCell,
+    is_64bits: BitCell,
+
+    boundary: UnlimitedCell,
+    aux1: U64OnU8Cell,
+    aux2: U64OnU8Cell,
+
+    lookup_pow: PowTableLookupCell,
+    lookup_stack_read: MTableLookupCell,
+    lookup_stack_write: MTableLookupCell,
+}
+
+pub struct UnaryConfigBuilder {}
+
+impl<F: FieldExt> EventTableOpcodeConfigBuilder<F> for UnaryConfigBuilder {
+    fn configure(
+        common: &mut EventTableCellAllocator<F>,
+        constraint_builder: &mut ConstraintBuilder<F>,
+    ) -> Box<dyn EventTableOpcodeConfig<F>> {
+        let operand = common.alloc_u64();
+        let result = common.alloc_u64();
+        let operand_is_zero = common.alloc_bit_value();
+        let operand_inv = common.alloc_unlimited_value();
+
+        let boundary = common.alloc_unlimited_value();
+        let aux1 = common.alloc_u64_on_u8();
+        let aux2 = common.alloc_u64_on_u8();
+
+        let is_ctz = common.alloc_bit_value();
+        let is_clz = common.alloc_bit_value();
+        let is_popcnt = common.alloc_bit_value();
+        let is_64bits = common.alloc_bit_value();
+
+        let lookup_pow = common.alloc_pow_table_lookup();
+        let lookup_stack_read = common.alloc_mtable_lookup();
+        let lookup_stack_write = common.alloc_mtable_lookup();
+
+        constraint_builder.push(
+            "op_unary: selector",
+            Box::new(move |meta| {
+                vec![
+                    (is_ctz.expr(meta) + is_clz.expr(meta) + is_popcnt.expr(meta)
+                        - constant_from!(1)),
+                ]
+            }),
+        );
+
+        constraint_builder.push(
+            "op_unary: zero cond",
+            Box::new(move |meta| {
+                vec![
+                    operand.expr(meta) * operand_inv.expr(meta) - constant_from!(1)
+                        + operand_is_zero.expr(meta),
+                ]
+            }),
+        );
+
+        constraint_builder.push(
+            "op_unary: clz",
+            Box::new(move |meta| {
+                let bits = constant_from!(32) + constant_from!(32) * is_64bits.expr(meta);
+
+                let is_32bits = constant_from!(1) - is_64bits.expr(meta);
+                let operand_is_not_zero = constant_from!(1) - operand_is_zero.expr(meta);
+
+                vec![
+                    operand_is_zero.expr(meta)
```

Comment:
just `32 + is_64bits.expr(meta) * 32`, better to keep expression simple unless it is too confusion.